yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -105,6 +102,16 @@ private[spark] class Client(
     // Finally, submit and monitor the application
     logInfo(s"Submitting application ${appId.getId} to ResourceManager")
     yarnClient.submitApplication(appContext)
+
+    // In YARN Cluster mode, AM is not killed when the client is terminated by default.
+    // But if spark.yarn.force.shutdown.am is set true, AM is force shutdown.
+    if (isClusterMode && sparkConf.getBoolean("spark.yarn.am.force.shutdown", false)) {
+      val shutdownHook = new Runnable {
+        override def run() { yarnClient.killApplication(appId) }
+      }
+      ShutdownHookManager.get().addShutdownHook(shutdownHook, 0)
```

Comment:
It seems hacky to put all this cleanup logic into shutdown hooks; can't this be in `stop()`?